File: go/ec2/ReleaseAddress/ReleaseAddress.go
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
// snippet-start:[ec2.go.release_address]
package main

// snippet-start:[ec2.go.release_address.imports]
import (
	"flag"
	"fmt"

	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/ec2"
	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
)
// snippet-end:[ec2.go.release_address.imports]

// ClearAddress releases an Elastic IP address.
// If the address is associated with an Amazon EC2 instance, the association is removed.
// Inputs:
//     svc is an Amazon EC2 service client
//     allocationID is the ID of the Elastic IP address
// Output:
//     If success, nil
//     Otherwise, an error from the call to ReleaseAddress
func ClearAddress(svc ec2iface.EC2API, allocationID *string) error {
	// snippet-start:[ec2.go.release_address.call]
	_, err := svc.ReleaseAddress(&ec2.ReleaseAddressInput{
		AllocationId: allocationID,
	})
	// snippet-end:[ec2.go.release_address.call]

	return err
}

func main() {
	// snippet-start:[ec2.go.release_address.args]
	allocationID := flag.String("a", "", "The ID of an allocated address")
	flag.Parse()

	if *allocationID == "" {
		fmt.Println("You must supply the ID of an allocated address (-a ALLOCATION-ID)")
		return
	}
	// snippet-end:[ec2.go.release_address.args]

	// snippet-start:[ec2.go.release_address.session]
	sess := session.Must(session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
	}))

	svc := ec2.New(sess)
	// snippet-end:[ec2.go.release_address.session]

	err := ClearAddress(svc, allocationID)
	if err != nil {
		fmt.Println("Got an error releasing address:")
		fmt.Println(err)
		return
	}

	fmt.Println("Released allocated address")
}
// snippet-end:[ec2.go.release_address]

File: java/example_code/ec2/src/main/java/aws/example/ec2/ReleaseAddress.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
package aws.example.ec2;

import com.amazonaws.services.ec2.AmazonEC2;
import com.amazonaws.services.ec2.AmazonEC2ClientBuilder;
import com.amazonaws.services.ec2.model.ReleaseAddressRequest;
import com.amazonaws.services.ec2.model.ReleaseAddressResult;

/**
 * Releases an elastic IP address
 */
public class ReleaseAddress {
    public static void main(String[] args) {
        final String USAGE = "To run this example, supply an allocation ID.\n" +
                "Ex: ReleaseAddress <allocation_id>\n";

        if (args.length != 1) {
            System.out.println(USAGE);
            System.exit(1);
        }

        String alloc_id = args[0];

        final AmazonEC2 ec2 = AmazonEC2ClientBuilder.defaultClient();

        ReleaseAddressRequest request = new ReleaseAddressRequest()
                .withAllocationId(alloc_id);

        ReleaseAddressResult response = ec2.releaseAddress(request);

        System.out.printf(
                "Successfully released elastic IP address %s", alloc_id);
    }
}
